protomap/proto.go

```diff
@@ -700,6 +726,137 @@ func protoFromPathsInternal(p proto.Message, vals map[*gpb.Path]any, valPrefix,
 	return nil
 }
 
+// createListField creates the entries of the repeated field fd within the protobuf message m, mapping the values within the val map.
+// valPrefix specifies the prefix to be applied to the paths within the vals map, protoPrefix specifies the prefix for the protobuf
+// message (if it is not the root), and fieldPath specifies the path to the field that is being mapped. ignoreExtras indicates whether
+// extra paths that do not exist in the message should be treated as errors.
+func createListField(m proto.Message, fd protoreflect.FieldDescriptor, fieldPath, valPrefix, protoPrefix *gpb.Path, vals map[*gpb.Path]any, ignoreExtras bool) (protoreflect.Value, error) {
+	keys := []map[string]string{}
+	keyPaths := []*gpb.Path{}
+
+	// We need to identify the keys that are within the list, as well as the data tree paths
+	// that they correspond to. We walk through the supplied values to determine which to process.
+	for p := range vals {
+		// Make the paths within the vals map absolute according to the supplied valPrefix.
+		absPath := &gpb.Path{
+			Elem: append(append([]*gpb.PathElem{}, valPrefix.Elem...), p.Elem...),
+		}
+		// Since the fieldPath is a schema path, then we need to compare just schema paths
+		// to avoid comparing the keys.
+		if !util.PathMatchesPathElemPrefix(schemaPath(absPath), schemaPath(fieldPath)) {
+			continue
+		}
+		// The key of the list is in the last element of the absolute path in the values map (the values
+		// map MUST contain data tree paths, since it is telling us list values to unmarshal).
+		k := absPath.Elem[len(fieldPath.Elem)-1]
+		// If the last element doesn't have a key, then we have not correctly found the list.
+		if k.Key == nil {
+			return protoreflect.Value{}, fmt.Errorf("invalid list schema field path does not have a list at the end %s", fieldPath)
```

Comment:
reword: invalid list data field path: does not have key values populated.

Also should the check be len(k.Key) == 0?

Another question is for unkeyed lists, since they can't be marshalled to a scalar, we can choose to not support them for now? If so, add a test on unkeyed lists that checks for an error.

Reply:
The protobuf translation itself doesn't handle unkeyed lists -- so they are never expected to show up here (it means that someone has created some proto with the annotations that they are somehow expecting to work, which seems unlikely).

Reworded the error and updated the check. I'm not sure how to add a test for the unkeyed list case, because we simply don't support them in the input proto, what does the test case that you're proposing look like? (We do not test this function with synthetic inputs because it's very difficult to create the right structures inside the protoreflect library that is a dependency).

Reply:
Not handling it sounds good to me. I didn't consider that it's not supported by the input proto. I'm happy with handling that when and if this becomes a use case.